### ISSUE TITLE
Unable to Sign Message using ethers.js Signer in mobile browser onAndroid

### ISSUE THREAD

**Issue by equalogix** [MetaMask/metamask-sdk#261]
**Describe the bug**
I'm using Sign Message to describe the bug, but this applies to any use of the signer, like for example sendTransaction or the use of Contract object to interact with smart contracts.

I am using `zkSync-web3` `Web3Provider` but this is a wrapper around ethers.js v5.7 `Web3Provider`. With respect to `signMessage` this should make no difference.

Basically signing a message does not work when using the Chrome browser on Android. It works fine using the desktop browser and connecting using the browser extension, and it also works using the desktop browser when connecting using the Mobile wallet (QR-code) (iOS and Android). It also works on iOS using Safari but the dApp doesn't always switch to the wallet, but if you manually switch the pop-up to sign the message is shown.

On Android everything appears to work correctly on the dApp side, with the SDK in Debug mode it shows the correct entries in the log, and the communications are appearing when inspecting the socket connection in the network tab.

When one first click the sign button no errors are displayed, and one can click the sign button several times, all without errors. But at some point the following error is returned as the result of  a click:
`App.js:20 Error: MetaMask is not connected/installed, please call eth_requestAccounts to connect first.`
Oddly the error does not occur after a specific timeout, at times it appears immediately after the second click of the button and other times it takes many clicks before it appears.

**To Reproduce**
Steps to reproduce the behaviour:
1.Using the code below see `Additional context`, on an Android device using the Chrome browser and Metamask Mobile app connect to the dApp.
2. Once connected, click the sign message button.
3. Nothing Happens, try again on iOS or desktop and the message will be signed as expected.

**Expected behaviour**
Message is signed when you call `signMessage`

**Screenshots**
NA

** Provide environment information**
- Platform Web React
- MetaMask SDK Version: `"@metamask/sdk-react": "^0.5.6"`
- MetaMask Mobile Version: V7.3.1 Android, V4iOS
- Browser (if relevant): google Chrome

**Additional context**
Here is the code:
File: MMSDKConnector.js
```js
import React, {
  createContext,
  useContext,
  useEffect,
  useState,
  useRef,
} from "react";
import { useSDK } from '@metamask/sdk-react';
import { Web3Provider, Provider } from "zksync-web3";

export const WalletContext = createContext(null);

export const WalletClient = ({children}) => {
    const [account, setAccount] = useState();
    const [signer, setSigner] = useState();
    const zkProviderRef = useRef();
    const {sdk, connected, connecting, provider, chainId} = useSDK();

    useEffect(() => {
      if(provider) {
        zkProviderRef.current = new Web3Provider(provider);
      }
    },[provider])

    const handleConnect = () => {
      console.log("connectizing sdk", sdk)
      sdk?.connect()
        .then( accts => {
          console.log("connected sdk", accts)
          setAccount(accts?.[0]);
        })
        .catch(err => {
          console.warn(`failed to connect..`, err);
        })
    }

    const handleDisconnect = () => {
      console.log("disconnectionizing sdk")
      sdk?.terminate();
      setAccount();
      setSigner();
      window.localStorage.clear();
      console.log("disconnectedized");
    }

    return (
      <WalletContext.Provider
        value={{
          account, setAccount,
          chainId,
          // providerRef,
          zkProviderRef,
          signer, setSigner,
          handleConnect, handleDisconnect, connected, connecting
        }}
      >
        {children}
      </WalletContext.Provider>
    )
  }

export const useAccount = () => {
  const { account, setAccount } = useContext(WalletContext); 
  return { account, setAccount }
}

export const useSigner = () => {
  const { signer, setSigner } = useContext(WalletContext); 
  return { signer, setSigner }
}

export const useChainId = () => {
  const { chainId, setChainId } = useContext(WalletContext); 
  return { chainId, setChainId }
}

export const useZkProvider = () => {
  const { zkProviderRef } = useContext(WalletContext); 
  const { current: zkProvider } = zkProviderRef;
  return { zkProvider }
}

export const useConnector = () => {
  const { 
    handleConnect, 
    handleDisconnect,
    connected,
    connecting
  } = useContext(WalletContext);
  return { handleConnect,handleDisconnect, connected, connecting }
}  

const styles = {
  button: {
    color: "#000",
    background: "#bbb",
    padding: 8,
    width: 120
  },
  widget: {
    display: "flex",
    justifyContent: "center",
    alignContent: "center",
    alignItems: "center",
    flexDirection: "column",
    width: "300",
    padding: 12,
    background: "#444",
    color: "#DDD",
    border: "solid #ddd 1px",
    borderRadius: 4
  }
}

export const MMConnector = () => {
  const { account } = useAccount();
  const { zkProvider } = useZkProvider();
  const { setSigner } = useSigner();
  const { handleConnect, handleDisconnect, connected, connecting } = useConnector();
  const chainIdInt = 280;

  useEffect(()=>{ 
    if(account && zkProvider) {
      setSigner(zkProvider.getSigner())
    }
  }, [account, zkProvider]);

  return (
    <>
    { connected ? (
      <>
      <button 
        style={styles.button}
        onClick={() => {handleDisconnect()}}
      >
        Disconnect
      </button>
      </>
    ) : ( connecting ? (
      <button 
        style={styles.button}
        onClick={() => {handleDisconnect()}}
      >
        Connecting...
      </button>
    ) : (
      <button 
        onClick={handleConnect}
        style={styles.button}
      >
        Connect
      </button>
    ))}
    </>
  )
}

export const WalletWidget = () => {
  const { account } = useAccount();
  const { chainId } = useChainId();
  
  return (
    <div style={styles.widget}>
      <h3>Connect Wallet</h3>
        <>
        { account && (
          <>
            <div>Connected with: MetaMaskSDK</div>
            <div>On Chain: {parseInt(chainId)} ({chainId})</div>
          </>
        )}
        <MMConnector />
        </>

    </div>
  )
}
```

File: App.js
```js
import React, { useState, useRef } from 'react';
import { WalletClient, WalletWidget } from "./MMSDKConnector";
import { MetaMaskProvider } from '@metamask/sdk-react';
import { useSigner } from "./MMSDKConnector";

export const SignMessageButton = (props) => {
  const {message} = props;
  const {signer} = useSigner(); 
  const [signedMsg, setSignedMsg] = useState();
  const timeoutRef = useRef();

  const signMessage = () => {
    if(!signer) return
    signer.signMessage(message)
      .then( x => { 
        console.log("signed msg:", x)
        setSignedMsg(x)
        timeoutRef.current = setTimeout( () => {setSignedMsg()}, 3000)
      })
      .catch( err => { console.error(err)})
  }

  return (
    <>
    {signer && <button onClick={signMessage}>Sign Message</button>}
    {signedMsg && <div style={{width:300, overflow:"clip"}}>{signedMsg}</div>}
    </>
  )
}


export default function App() {
  return (
    <MetaMaskProvider
      debug 
      sdkOptions={{
        logging:{
            developerMode: true,
          },
          checkInstallationImmediately: false, // This will automatically connect to MetaMask on page load
          checkInstallationOnAllCalls: false,
          dappMetadata: {
            name: "Demo React App",
            url: window.location.host,
          }
      }}
    >
    <WalletClient>
      <div style={{
        display: "flex",
        flexDirection: "column",
        justifyContent: "center",
        alignItems: "center",
        margin: 16,
        padding: 16,
        background: "#333",
        borderRadius: 16
      }}>
        <WalletWidget/>
        <div style={{
          background: "#ccc",
          padding:16,
          margin: 16,
          borderRadius: 16,
          textAlign: "center"
        }}>
        <SignMessageButton message="Test Message!" />
        </div>
      </div>
    </WalletClient>
    </MetaMaskProvider>
  )
}
```

**Comment by andreahaku** [comment]
Hi @equalogix, we'll try to replicate the issue using your code and we'll get back to you.

The team is off this week and that's why we are a little slow on answering, testing and fixing. But from Monday everything will get back to normal.

**Comment by abretonc7s** [comment]
Hi @equalogix ,

Sorry for the slow reply and thank you for providing us with additional information.  I will investigate more on the Contract integration and see how it differs from calling sendTransaction directly.
I assume you your workaround is fixing the issue? Can we close the ticket?

**Comment by christopherferreira9** [comment]
Hi @equalogix ! We were still unable to get on top of this issue but it hasn't been forgotten, we will try to reproduce as soon as possible.

**Comment by andreahaku** [comment]
Hi @equalogix we are setting up to reproduce the issue and fix it. Sorry for the delay. We'll keep you updated.

**Comment by abretonc7s** [comment]
Hi @equalogix ,

The issue is due to ethers library first sending an eth_accounts request before the sign request. Since eth_accounts doesn't triggers a deeplink on  sdk < 0.7.0 it seems to hangs. We are now implementing "read-only" calls with v0.7.0+ which will prevent this situation. We will revert back to you as soon as we have updated our documentation for settings up read only calls.

**Comment by christopherferreira9** [comment]
Hi @equalogix ! Were you able to try the latest SDK version? This should be fixed from `0.7.1`.
